gpsrenda/widgets/widgets.py

```diff
@@ -0,0 +1,192 @@
+        gauge_class = STYLE_TABLE[style]
+        gauge = gauge_class(x, y, label="{val:.0f}", dummy_label="999", caption="bpm",
+                            data_range=[(100, [0.2, 0.0, 0.6]),
+                                        (200, [0.8, 0.0, 0.0])])
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.heart_rate(t)
+        self.gauge.render(context, value)
+
+
+class PowerWidget:
+    def __init__(self, data_source, x, y, style='hbar'):
+        self.data_source = data_source
+        gauge_class = STYLE_TABLE[style]
+        gauge = gauge_class(x, y, label="{val:.0f}", dummy_label="999", caption="W",
+                            data_range=[(0, (0, 0.6, 0)),
+                                        (300, (0.2, 0.6, 0.0)),
+                                        (800, (0.8, 0.0, 0.0))])
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.power(t)
+        self.gauge.render(context, value)
+
+
+class TemperatureWidget:
+    def __init__(self, data_source, x, y, style='vbar', units='metric'):
+        self.data_source = data_source
+        self.units = units
+        gauge_class = STYLE_TABLE[style]
+        suffix = '°F' if units == 'imperial' else '°C'
+
+        colors = [(5, (0.6, 0, 0)),
+                  (20, (0.0, 0.6, 0.0)),
+                  (35, (0.0, 0.0, 1.0))]
+
+        if units == 'imperial':
+            for idx, (val, rgb) in enumerate(colors):
+                colors[idx] = (c_to_f(val), rgb)
+
+        gauge = gauge_class(x, y, label="{val:.0f}"+suffix, dummy_label="0.0",
+                            data_range=colors)
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.temperature(t)
+        if self.units == 'imperial':
+            value = c_to_f(value)
+        self.gauge.render(context, value)
+
+
+class DistanceWidget:
+    def __init__(self, data_source, x, y, w, style='hbar', units='metric'):
+        self.data_source = data_source
+        self.units = units
+        gauge_class = STYLE_TABLE[style]
+        suffix = 'mi' if units == 'imperial' else 'km'
+
+        total_distance = data_source.fields['distance'][-1][1]
+
+        if units == 'imperial':
+            total_distance = km_to_mi(total_distance) / 1000
+
+            colors = [(0, [0.75, 0.75, 0.75]),
+                      (total_distance, [0.75, 0.75, 0.75])]
+
+
+        gauge = gauge_class(x, y, w=w, label="{val:.1f}", dummy_label="0.0", caption=f" / {total_distance:.1f} {suffix}",
+                            data_range=colors)
+        self.gauge = gauge
+
+    def render(self, context, t):
+        value = self.data_source.distance(t) / 1000
+        if self.units == 'imperial':
+            value = km_to_mi(value)
+        self.gauge.render(context, value)
+
+
+class MapWidget:
+    PRIVACY_RANGE = 2000 #m
+
+    def __init__(self, data_source, x, y, h, style='map', privacy=True):
```

Comment:
maybe pass in `privacy_range` with a default of `None`, rather than hardcoding `PRIVACY_RANGE`?

Reply:
Strava has this feature and uses a fixed distance. I figure no one really cares exactly what the range is if it's reasonable. If you're the only house in a 2km radius, people are gonna figure out where you live based on your rides.